### PR TITLE
Get task

### LINKED ISSUE #3: Action Server: Transit Action
The Transit action is a kind of task which involves changing of a character or party's location

- [X] extending ontologies
- [x] code superstructure for actions (probably similar to content describer system)
- [x] endpoint for user to create a Transit task
- [x] endpoint for user to finish the task

## Notes

See [this thread](https://forum.solidproject.org/t/modelling-changes-to-graphs-in-owl-rdf/3872) on the initial design points

The task is initially planned to use an N3 formula (a node which points to a graph) to describe the changes to world/character state in the `endState` or consequences of the task. I'm not sure that the clientside will be able to handle it, so we may need to revert to something which can be implemented solely in Turtle

### PR DESCRIPTION
**Pull request by calummackervoy**
Closes #3 

:partying_face: 

* Instantiating a DataSetItem from an existing dataset, TaskActors can be instantiated from an existing source too
* AbstractMUDController providing serializeToTurtle for convenience
* TaskController GET. Gets a TaskActor with the existing task resource. Checks if complete, and if it is effectuates the task consequences and returns the complete task. If it isn't, just returns the task
* refactored Task dataset management code so that there's a separation of concerns
* the Time vocab class has some functions to convert Time resources to and fro Java objects. Refactored StadiumDescriber to use them

### CODE REVIEW

**Comment by calummackervoy** [comment]
@MattTennison do you know what's causing the build to fail ?

**Comment by calummackervoy** [comment]
From the build error logs

```
#13 6.441 [INFO] Changes detected - recompiling the module!
11472
#13 6.445 [INFO] Compiling 29 source files to /app/target/classes
11473
#13 7.500 [INFO] -------------------------------------------------------------
11474
#13 7.502 [ERROR] COMPILATION ERROR : 
11475
#13 7.502 [INFO] -------------------------------------------------------------
11476
#13 7.503 [ERROR] /app/src/main/java/com/mackervoy/calum/mud/behaviour/task/AbstractTaskActor.java:[31,45] package org.eclipse.jdt.internal.compiler.ast does not exist
11477
#13 7.507 [ERROR] /app/src/main/java/com/mackervoy/calum/mud/content/ContentController.java:[57,65] cannot find symbol
11478
#13 7.507   symbol:   method modelToTurtle(org.apache.jena.rdf.model.Model)
11479
#13 7.549   location: class com.mackervoy.calum.mud.content.ContentController
11480
#13 7.549 [INFO] 2 errors 
11481
#13 7.549 [INFO] -------------------------------------------------------------
11482
#13 7.549 [INFO] ------------------------------------------------------------------------
11483
#13 7.549 [INFO] BUILD FAILURE
11484
#13 7.549 [INFO] ------------------------------------------------------------------------
11485
#13 7.549 [INFO] Total time:  6.037 s
11486
#13 7.549 [INFO] Finished at: 2021-04-19T16:52:18Z
11487
#13 7.549 [INFO] ------------------------------------------------------------------------
11488
#13 7.555 [ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project mud: Compilation failure: Compilation failure: 
11489
#13 7.573 [ERROR] /app/src/main/java/com/mackervoy/calum/mud/behaviour/task/AbstractTaskActor.java:[31,45] package org.eclipse.jdt.internal.compiler.ast does not exist
11490
#13 7.573 [ERROR] /app/src/main/java/com/mackervoy/calum/mud/content/ContentController.java:[57,65] cannot find symbol
11491
#13 7.573 [ERROR]   symbol:   method modelToTurtle(org.apache.jena.rdf.model.Model)
11492
#13 7.573 [ERROR]   location: class com.mackervoy.calum.mud.content.ContentController
11493
#13 7.573 [ERROR] -> [Help 1]
11494
#13 7.573 [ERROR] 
11495
#13 7.573 [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
11496
#13 7.573 [ERROR] Re-run Maven using the -X switch to enable full debug logging.
11497
#13 7.573 [ERROR] 
11498
#13 7.573 [ERROR] For more information about the errors and possible solutions, please read the following articles:
11499
#13 7.573 [ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
11500
#13 ERROR: executor failed running [/bin/sh -c mvn package]: exit code: 1
11501
------
11502
 > [maven_deps 7/7] RUN mvn package:
11503
------
11504
error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c mvn package]: exit code: 1
11505
Error: buildx call failed with: error: failed to solve: rpc error: code = Unknown desc = executor failed running [/bin/sh -c mvn package]: exit code: 1
```

Not sure I get what's going on there 😅

**Comment by calummackervoy** [comment]
@MattTennison are you able to pull these changes and try setting up a server your side? I've got no problem running it on Tomcat in Eclipse on mine

**Comment by MattTennison** [comment]
Taking a look at it now. Sounds like a missing Maven dependency for this `org.eclipse.jdt.internal.compiler.ast` package (which I assume Eclipse will just magically make available in its environment)

**Comment by MattTennison** [comment]
Pushed up a branch with some build fixes for you to review @calummackervoy :smile: - https://github.com/Multi-User-Domain/mud-jena/pull/38

File: src/main/java/com/mackervoy/calum/mud/TDBStore.java [hunk 1]
```diff
@@ -23,4 +29,70 @@ public final static boolean inUseLocation(String fileLocation) {
...
+	public final static DatasetItem getDatasetItem(String uriOrFilePath) {
+		
+		// convert a URI to a file system path
+		String filePath = getFilePathFromUri(uriOrFilePath).orElse(uriOrFilePath);
+		
+		if(!TDBStore.inUseLocation(new File(getFullFilePath(filePath)))) {
+			throw new NotFoundException("The given dataset is not in use!");
+		}
+		
+		String[] splitString = filePath.split("/");
```
Comment (MattTennison) [inline src/main/java/com/mackervoy/calum/mud/TDBStore.java]: This sorta thing feels like we should be delegating it to a third party library (or wrapping it), to keep our codebase focused on the business logic in the app - e.g. https://commons.apache.org/proper/commons-io/javadocs/api-1.4/org/apache/commons/io/FilenameUtils.html#getName(java.lang.String)

File: src/main/java/com/mackervoy/calum/mud/behaviour/task/AbstractTaskActor.java [hunk 2]
```diff
@@ -95,18 +122,74 @@ protected Resource getFirstResourceMatchingType(Model request, Resource type) {
...
+		Resource end = this.task.getPropertyResourceValue(Time.hasEnd);
+		if(end == null) return Optional.empty();
 		
-		Dataset dataset = TDB2Factory.connectDataset(this.taskDatasetItem.getFileLocation());
-		try {
-			dataset.begin(ReadWrite.WRITE);
-			Model out = dataset.getDefaultModel();
-			out.add(this.model);
-			out.commit();
+		return Optional.of(Time.instantToLocalDateTime(end));
```
Comment (MattTennison) [inline src/main/java/com/mackervoy/calum/mud/behaviour/task/AbstractTaskActor.java]: Using local time will tie us to the local timezone of the server, and it doesn't look like it'll expose that timezone information in the outputted string (meaning we couldn't tell the user the time in their timezone, or what timezone it's in without also exposing the timezone of the server). 

Can we use UTC on the server, and then convert it out into a users timezone on the front end instead?

Comment (calummackervoy) [reply src/main/java/com/mackervoy/calum/mud/behaviour/task/AbstractTaskActor.java]: I think using UTC as a standard makes sense :+1: 

Also made me think of this one https://github.com/Multi-User-Domain/vocab/issues/8

File: src/main/java/com/mackervoy/calum/mud/TDBStore.java (hunk 1 above)
Comment (calummackervoy) [reply src/main/java/com/mackervoy/calum/mud/TDBStore.java]: Yeah spot on :+1:

Comment (calummackervoy) [reply src/main/java/com/mackervoy/calum/mud/TDBStore.java]: Updated this MR to at least use the `FilenameUtils` in string manipulation and such in `TDBStore` and `AbstractFileStorageWrapper`... does this cover it?

In terms of the general feature.. I'm not sure how much of the datastore management the platform should be responsible for and how much we can outsource: #32 

In terms of the purpose of the repo I think this repository is responsible for providing a Java library to build a MUD application, but that a particular server with custom features would use + extend it .. I'm not sure where the line is - I opened this about it if you have thoughts #34

**Pull request merged**